content/howto/bash.md

```diff
@@ -0,0 +1,969 @@
+
+```bash
+$ ls folder > file.txt 2> errors
+$ cat errors
+ls: cannot access 'folder': No such file or directory
+```
+
+zapis `2>` wskazuje, że mamy przekazywać do strumienia nr 2 którym domyślnie jest STDERR. Nie jest to zbyt oczywiste, ale taki bash już niestety jest.
+
+znak `>` nadpisuje plik. Jeżeli chcemy dodawać nowy content do pliku możemy skorzystać z `>>`:
+
+```bash
+$ echo 111 >> file.txt
+$ echo 222 >> file.txt
+$ echo 333 >> file.txt
+$ cat file.txt
+111
+222
+333
+```
+
+Możemy też w drugą stronę - przekazywać strumień z innego pliku do programu (STDIN) np.
+
+```bash
+$ cat < file.txt
+111
+222
+333
+```
+
+Osobiście z operatora `<` rzadko stosuję bo programy zwykle mają odczyt z plików w postaci parametrów.
+
+### Wirtualne pliki
+
+W systemie linux mamy kilka wirtualnych plików:
+
+* `/dev/null` - pustka, wysłanie do niego źródeł kasuje je bezpowrotnie. Stosuje się gdy chcesz by nie drukowało na ekranie
+* `/dev/random` - wysyła losowe wartości, dobre do nadpisania plików (skasowany plik tak naprawdę nie kasuje wartośći, tylko oznacza ją do ponownego zapisu)
+* `/dev/urandom` - to samo co `/dev/random` ale robi to szybciej ale mniej bezpieczniej (szczegóły sobie sam doczytaj)
+* `/dev/zero` - wysyła same zera (w sense liczbowym, a nie znakiem 0 który ma wartość 48, tablica ascii)
+
+## Pipe 
+
+Jak już poznaliśmy jak wrzucać strumień z ekranu do pliku to możemy poznać pipe czyli po polsku rury.
+Operator `|` (nazywanym operatorem pipe) pozwala nam na uruchamianie serię komend 
+gdzie kolejna komenda pobiera z strumienia poprzedniej komendy dane. Brzmi coś obco? To może przykładzik:
+
+```bash
+$ ls folder | wc -l
+94
+```
+
+1. Wyswietlamy wszystki pliki w folderze `folder` 
+2. Strumień z komendy `ls` jest pobierany przez komendę `wc -l` jako wejście
+
+Możemy robić ciekawsze rzeczy jak np. sprawdzenie binarnego pliku PNG. Do tego potrzebujemy dwóch nowych komend:
+
+* `head` - wyświetla kilka pierwszych linii z wejścia.
+* `hexdump` - wyświetla binarne dane w postaci hex - każdy bajt jest pokazany jako dwa liczby 16-tkowe.
+
+Jedziem z przykładem:
+
+```bash
+$ cat plik.png | hexdump -C | head -n4
+00000000  89 50 4e 47 0d 0a 1a 0a  00 00 00 0d 49 48 44 52  |.PNG........IHDR|
+00000010  00 00 01 e4 00 00 03 5c  08 02 00 00 00 01 ca 4b  |.......\.......K|
+00000020  9f 00 00 00 03 73 42 49  54 08 08 08 db e1 4f e0  |.....sBIT.....O.|
+00000030  00 00 20 00 49 44 41 54  78 9c ec dd 77 58 14 c7  |.. .IDATx...wX..|
+```
+
+1. wyswietlismy plik binarny
+2. `hexdump -C` pokazał formę 16-tkową oraz znaki ASCII po prawej stronie
+3. `head -n4` pokazał 4 linijki wyścia z komendy `hexdump -C`
+
+To by było na tyle z podstaw i zaawansowanych rzeczy - jest oczywiście sporo innych rzeczy 
+ale to wystarczy by spokojnie się poruszać i wykorzystywac pełną moc shella w linuxie :)
+
+# 4. Podstawowe komendy do pracy z plikami/folderami ## {: #commands}
+
+## ls
+
+ls = **L**is**T** Wyświetla folder. Po prostu.
```

Comment:
o co chodzi z tym  **L**is**T** ?

Reply:
nie wyświetla folderu tylko listę plików w danym katalogu

Reply:
podkreślam które literki oznaczają za skrót